File: native/FileFlow.Core/Executor.cs
namespace FileFlow.Core;

public static class Executor
{
    public static string Execute(OperationPlan plan, IProgress<(int Current, int Total, string Name)>? progress = null)
    {
        if (plan.Items.Count == 0) throw new SafetyException("적용할 파일이 없습니다.");
        var completed = new List<CompletedItem>();
        var createdDestination = false;
        try
        {
            if (plan.Kind == OperationKind.Rename)
                completed.AddRange(ExecuteRename(plan, progress));
            else
            {
                var destinationRoot = plan.DestinationRoot ?? throw new SafetyException("대상 폴더 정보가 없습니다.");
                if (!Directory.Exists(destinationRoot))
                {
                    Directory.CreateDirectory(destinationRoot);
                    createdDestination = true;
                }
                for (var index = 0; index < plan.Items.Count; index++)
                {
                    var item = plan.Items[index];
                    VerifySource(item);
                    VerifyBoundary(plan, item);
                    if (File.Exists(item.Destination) || Directory.Exists(item.Destination))
                        throw new SafetyException($"대상 파일이 이미 존재합니다: {Path.GetFileName(item.Destination)}");
                    switch (item.Action)
                    {
                        case FileAction.Copy: File.Copy(item.Source, item.Destination, false); break;
                        case FileAction.Move: File.Move(item.Source, item.Destination, false); break;
                        default: throw new SafetyException("지원하지 않는 작업입니다.");
                    }
                    completed.Add(new(item.Source, item.Destination, item.Action, FileFingerprint.FromPath(item.Destination)));
                    progress?.Report((index + 1, plan.Items.Count, Path.GetFileName(item.Destination)));
                }
            }
            var deletedDirectories = plan.DeleteEmpty && plan.SourceRoot is not null ? DeleteEmptySubdirectories(plan.SourceRoot) : [];
            var journal = new OperationJournal
            {
                Id = plan.Id,
                Status = "success",
                Kind = plan.Kind,
                Plan = plan,
                Completed = completed,
                DeletedDirectories = deletedDirectories,
                CreatedDestination = createdDestination
            };
            return JournalStore.Write(journal);
        }
        catch (Exception error)
        {
            var rollbackErrors = plan.Kind == OperationKind.Flatten ? RollbackFlatten(completed) : [];
            if (createdDestination && plan.DestinationRoot is not null)
            {
                try { Directory.Delete(plan.DestinationRoot); } catch (IOException) { }
            }
            JournalStore.Write(new()
            {
                Id = plan.Id,
                Status = "failed",
                Kind = plan.Kind,
                Plan = plan,
                Completed = completed,
                CreatedDestination = createdDestination,
                Error = error.Message,
                RollbackErrors = rollbackErrors
            });
            if (rollbackErrors.Count > 0)
                throw new InvalidOperationException($"{error.Message}\n자동 롤백 확인 필요: {string.Join("; ", rollbackErrors)}", error);
            throw;
        }
    }

    public static int UndoLatest(IProgress<(int Current, int Total, string Name)>? progress = null)
    {
        var journal = JournalStore.LoadLatest() ?? throw new SafetyException("실행 취소할 직전 작업이 없습니다.");
        var completed = journal.Completed;
        if (journal.Kind == OperationKind.Rename)
            UndoRename(completed);
        else
        {
            foreach (var item in completed)
            {
                if (!File.Exists(item.Destination) || !Matches(item.Destination, item.DestinationFingerprint))
                    throw new SafetyException($"Undo 대상이 변경되었습니다: {Path.GetFileName(item.Destination)}");
                if (item.Action == FileAction.Move && (File.Exists(item.Source) || Directory.Exists(item.Source)))
                    throw new SafetyException($"원래 위치가 이미 사용 중입니다: {item.Source}");
            }
            foreach (var directory in journal.DeletedDirectories.AsEnumerable().Reverse()) Directory.CreateDirectory(directory);
            for (var index = 0; index < completed.Count; index++)
            {
                var item = completed[completed.Count - 1 - index];
                if (item.Action == FileAction.Copy) File.Delete(item.Destination);
                else
                {
                    Directory.CreateDirectory(Path.GetDirectoryName(item.Source)!);
                    File.Move(item.Destination, item.Source, false);
                }
                progress?.Report((index + 1, completed.Count, Path.GetFileName(item.Source)));
            }
            if (journal.CreatedDestination && journal.Plan.DestinationRoot is not null)
            {
                try { Directory.Delete(journal.Plan.DestinationRoot); } catch (IOException) { }
            }
        }
        journal.UndoneAt = DateTimeOffset.UtcNow;
        journal.Status = "undone";
        JournalStore.Write(journal);
        return completed.Count;
    }

    private static IEnumerable<CompletedItem> ExecuteRename(OperationPlan plan, IProgress<(int, int, string)>? progress)
    {
        var active = plan.Items.Where(item => !item.Source.Equals(item.Destination, StringComparison.OrdinalIgnoreCase)).ToList();
        foreach (var item in active) { VerifySource(item); VerifyBoundary(plan, item); }
        var sourceSet = active.Select(item => item.Source).ToHashSet(StringComparer.OrdinalIgnoreCase);
        foreach (var item in active)
            if (File.Exists(item.Destination) && !sourceSet.Contains(item.Destination))
                throw new SafetyException($"변경 후 파일명이 이미 존재합니다: {Path.GetFileName(item.Destination)}");

        var staged = new List<(PlanItem Item, string Temporary)>();
        var completed = new List<(PlanItem Item, string Current)>();
        try
        {
            foreach (var item in active)
            {
                var temporary = Path.Combine(Path.GetDirectoryName(item.Source)!, $".fileflow-stage-{Guid.NewGuid():N}.tmp");
                File.Move(item.Source, temporary, false);
                staged.Add((item, temporary));
            }
            for (var index = 0; index < staged.Count; index++)
            {
                var entry = staged[index];
                File.Move(entry.Temporary, entry.Item.Destination, false);
                completed.Add((entry.Item, entry.Item.Destination));
                progress?.Report((index + 1, staged.Count, Path.GetFileName(entry.Item.Destination)));
            }
        }
        catch
        {
            var completedItems = completed.Select(entry => entry.Item).ToHashSet();
            var locations = staged.Select(entry => (entry.Item, completedItems.Contains(entry.Item) ? entry.Item.Destination : entry.Temporary)).ToList();
            RollbackRename(locations);
            throw;
        }
        return active.Select(item => new CompletedItem(item.Source, item.Destination, FileAction.Rename, FileFingerprint.FromPath(item.Destination))).ToList();
    }

    private static void UndoRename(List<CompletedItem> completed)
    {
        foreach (var item in completed)
            if (!File.Exists(item.Destination) || !Matches(item.Destination, item.DestinationFingerprint))
                throw new SafetyException($"Undo 대상이 변경되었습니다: {Path.GetFileName(item.Destination)}");
        var destinations = completed.Select(item => item.Destination).ToHashSet(StringComparer.OrdinalIgnoreCase);
        foreach (var item in completed)
            if ((File.Exists(item.Source) || Directory.Exists(item.Source)) && !destinations.Contains(item.Source))
                throw new SafetyException($"원래 파일명이 이미 사용 중입니다: {Path.GetFileName(item.Source)}");
        var parked = new List<(string Temporary, CompletedItem Item)>();
        try
        {
            foreach (var item in completed)
            {
                var temporary = Path.Combine(Path.GetDirectoryName(item.Destination)!, $".fileflow-undo-{Guid.NewGuid():N}.tmp");
                File.Move(item.Destination, temporary, false);
                parked.Add((temporary, item));
            }
        }
        catch
        {
            foreach (var entry in parked.AsEnumerable().Reverse())
                if (File.Exists(entry.Temporary) && !File.Exists(entry.Item.Destination)) File.Move(entry.Temporary, entry.Item.Destination, false);
            throw;
        }

        var restored = new List<(CompletedItem Item, string Current)>();
        try
        {
            foreach (var entry in parked)
            {
                File.Move(entry.Temporary, entry.Item.Source, false);
                restored.Add((entry.Item, entry.Item.Source));
            }
        }
        catch
        {
            var restoredSet = restored.Select(entry => entry.Item).ToHashSet();
            var current = parked.Select(entry => (entry.Item, restoredSet.Contains(entry.Item) ? entry.Item.Source : entry.Temporary)).ToList();
            var rollback = new List<(string Temporary, string Destination)>();
            foreach (var entry in current)
            {
                if (!File.Exists(entry.Item2)) continue;
                var temporary = Path.Combine(Path.GetDirectoryName(entry.Item.Destination)!, $".fileflow-undo-rollback-{Guid.NewGuid():N}.tmp");
                File.Move(entry.Item2, temporary, false);
                rollback.Add((temporary, entry.Item.Destination));
            }
            foreach (var entry in rollback) File.Move(entry.Temporary, entry.Destination, false);
            throw;
        }
    }

    private static void RollbackRename(IEnumerable<(PlanItem Item, string Current)> locations)
    {
        var parked = new List<(string Temporary, string Source)>();
        foreach (var (item, current) in locations)
        {
            if (!File.Exists(current)) continue;
            var temporary = Path.Combine(Path.GetDirectoryName(item.Source)!, $".fileflow-rollback-{Guid.NewGuid():N}.tmp");
            File.Move(current, temporary, false);
            parked.Add((temporary, item.Source));
        }
        foreach (var entry in parked)
        {
            if (File.Exists(entry.Source)) throw new IOException($"롤백 대상이 이미 존재합니다: {entry.Source}");
            File.Move(entry.Temporary, entry.Source, false);
        }
    }

    private static List<string> RollbackFlatten(IEnumerable<CompletedItem> completed)
    {
        var errors = new List<string>();
        foreach (var item in completed.Reverse())
        {
            try
            {
                if (!File.Exists(item.Destination) || !Matches(item.Destination, item.DestinationFingerprint))
                    throw new SafetyException($"롤백 대상 상태가 달라졌습니다: {item.Destination}");
                if (item.Action == FileAction.Copy) File.Delete(item.Destination);
                else
                {
                    if (File.Exists(item.Source)) throw new SafetyException($"롤백 원본 위치가 사용 중입니다: {item.Source}");
                    Directory.CreateDirectory(Path.GetDirectoryName(item.Source)!);
                    File.Move(item.Destination, item.Source, false);
                }
            }
            catch (Exception ex) { errors.Add(ex.Message); }
        }
        return errors;
    }

    private static void VerifySource(PlanItem item)
    {
        if (!File.Exists(item.Source) || Safety.ExclusionReason(item.Source) is not null)
            throw new SafetyException($"원본 파일이 없거나 안전하지 않습니다: {item.Source}");
        if (!Matches(item.Source, item.Fingerprint))
            throw new SafetyException($"미리보기 후 파일이 변경되었습니다: {Path.GetFileName(item.Source)}");
    }

    private static void VerifyBoundary(OperationPlan plan, PlanItem item)
    {
        if (plan.Kind == OperationKind.Flatten)
            Safety.ValidateDestination(item.Destination, plan.DestinationRoot ?? throw new SafetyException("대상 폴더 정보가 없습니다."));
        else
            Safety.ValidateDestination(item.Destination, Path.GetDirectoryName(item.Source)!);
    }

    private static bool Matches(string path, FileFingerprint fingerprint)
    {
        try { return FileFingerprint.FromPath(path) == fingerprint; }
        catch (IOException) { return false; }
        catch (UnauthorizedAccessException) { return false; }
    }

    private static List<string> DeleteEmptySubdirectories(string root)
    {
        var removed = new List<string>();
        var discovered = new List<string>();
        var stack = new Stack<string>();
        stack.Push(root);
        while (stack.Count > 0)
        {
            var current = stack.Pop();
            IEnumerable<string> children;
            try { children = Directory.EnumerateDirectories(current).ToArray(); }
            catch (Exception ex) when (ex is IOException or UnauthorizedAccessException) { continue; }
            foreach (var child in children)
            {
                try
                {
                    if ((File.GetAttributes(child) & FileAttributes.ReparsePoint) != 0) continue;
                    discovered.Add(child);
                    stack.Push(child);
                }
                catch (Exception ex) when (ex is IOException or UnauthorizedAccessException) { }
            }
        }
        foreach (var directory in discovered.OrderByDescending(path => path.Length))
        {
            try
            {
                Directory.Delete(directory, false);
                removed.Add(directory);
            }
            catch (IOException) { }
            catch (UnauthorizedAccessException) { }
        }
        return removed;
    }
}

File: src/fileflow_lite/core/executor.py
from __future__ import annotations

import os
import shutil
from pathlib import Path
from typing import Any, Callable
from uuid import uuid4

from .journal import load_latest, mark_undone, write_journal
from .models import FileFingerprint, OperationPlan, PlanItem
from .safety import SafetyError, is_within, validate_destination

Progress = Callable[[int, int, str], None]


def _fingerprint(path: Path) -> FileFingerprint:
    return FileFingerprint.from_path(path)


def _verify_source(item: PlanItem) -> None:
    path = item.source_path
    if not path.is_file() or path.is_symlink():
        raise SafetyError(f"원본 파일이 없거나 안전하지 않습니다: {path}")
    if _fingerprint(path) != item.fingerprint:
        raise SafetyError(f"미리보기 후 파일이 변경되었습니다: {path.name}")


def _verify_plan_boundary(plan: OperationPlan, item: PlanItem) -> None:
    if plan.kind == "flatten":
        if not plan.destination_root:
            raise SafetyError("대상 폴더 정보가 없습니다.")
        validate_destination(item.destination_path, Path(plan.destination_root))
    else:
        validate_destination(item.destination_path, item.source_path.parent)


def _delete_empty_subfolders(source_root: Path) -> list[str]:
    removed: list[str] = []
    for directory, _, _ in os.walk(source_root, topdown=False, followlinks=False):
        path = Path(directory)
        if path == source_root or path.is_symlink():
            continue
        try:
            path.rmdir()
            removed.append(str(path))
        except OSError:
            pass
    return removed


def _rollback_rename(locations: list[tuple[PlanItem, Path]]) -> None:
    parked: list[tuple[Path, Path]] = []
    for item, current in locations:
        if not current.exists():
            continue
        rollback_temp = item.source_path.parent / f".fileflow-rollback-{uuid4().hex}.tmp"
        os.replace(current, rollback_temp)
        parked.append((rollback_temp, item.source_path))
    for temp, source in parked:
        if source.exists():
            raise RuntimeError(f"롤백 대상이 이미 존재합니다: {source}")
        os.replace(temp, source)


def _execute_rename(plan: OperationPlan, progress: Progress | None) -> list[dict[str, Any]]:
    active = [i for i in plan.items if os.path.normcase(i.source) != os.path.normcase(i.destination)]
    for item in active:
        _verify_source(item)
        _verify_plan_boundary(plan, item)

    source_keys = {os.path.normcase(i.source) for i in active}
    for item in active:
        if item.destination_path.exists() and os.path.normcase(item.destination) not in source_keys:
            raise SafetyError(f"변경 후 파일명이 이미 존재합니다: {item.destination_path.name}")

    staged: list[tuple[PlanItem, Path]] = []
    try:
        for item in active:
            temp = item.source_path.parent / f".fileflow-stage-{uuid4().hex}.tmp"
            os.replace(item.source_path, temp)
            staged.append((item, temp))
        completed: list[tuple[PlanItem, Path]] = []
        for index, (item, temp) in enumerate(staged, 1):
            os.replace(temp, item.destination_path)
            completed.append((item, item.destination_path))
            if progress:
                progress(index, len(active), item.destination_path.name)
    except Exception:
        current_locations = []
        completed_sources = {id(item) for item, _ in locals().get("completed", [])}
        for item, temp in staged:
            current = item.destination_path if id(item) in completed_sources else temp
            current_locations.append((item, current))
        _rollback_rename(current_locations)
        raise

    return [
        {
            "source": item.source,
            "destination": item.destination,
            "action": "rename",
            "destination_fingerprint": _fingerprint(item.destination_path).__dict__,
        }
        for item in active
    ]


def execute_plan(plan: OperationPlan, progress: Progress | None = None) -> Path:
    if not plan.items:
        raise SafetyError("적용할 파일이 없습니다.")
    completed: list[dict[str, Any]] = []
    created_destination = False
    try:
        if plan.kind == "rename":
            completed = _execute_rename(plan, progress)
        else:
            destination_root = Path(plan.destination_root or "")
            if not destination_root.exists():
                destination_root.mkdir(parents=False)
                created_destination = True
            total = len(plan.items)
            for index, item in enumerate(plan.items, 1):
                _verify_source(item)
                _verify_plan_boundary(plan, item)
                destination = item.destination_path
                if destination.exists():
                    raise SafetyError(f"대상 파일이 이미 존재합니다: {destination.name}")
                if item.action == "copy":
                    shutil.copy2(item.source_path, destination)
                elif item.action == "move":
                    shutil.move(str(item.source_path), str(destination))
                else:
                    raise SafetyError("지원하지 않는 작업입니다.")
                completed.append(
                    {
                        "source": item.source,
                        "destination": item.destination,
                        "action": item.action,
                        "destination_fingerprint": _fingerprint(destination).__dict__,
                    }
                )
                if progress:
                    progress(index, total, destination.name)

        deleted_dirs: list[str] = []
        if plan.kind == "flatten" and plan.delete_empty and plan.source_root:
            deleted_dirs = _delete_empty_subfolders(Path(plan.source_root))
        payload = {
            "id": plan.id,
            "status": "success",
            "kind": plan.kind,
            "plan": plan.to_dict(),
            "completed": completed,
            "deleted_dirs": deleted_dirs,
            "created_destination": created_destination,
        }
        return write_journal(payload)
    except Exception as exc:
        rollback_errors: list[str] = []
        if plan.kind == "flatten":
            for entry in reversed(completed):
                source = Path(entry["source"])
                destination = Path(entry["destination"])
                try:
                    if not destination.exists() or not _same_fingerprint(
                        destination, entry["destination_fingerprint"]
                    ):
                        raise SafetyError(f"롤백 대상 상태가 달라졌습니다: {destination}")
                    if entry["action"] == "copy":
                        destination.unlink()
                    elif entry["action"] == "move":
                        if source.exists():
                            raise SafetyError(f"롤백 원본 위치가 사용 중입니다: {source}")
                        source.parent.mkdir(parents=True, exist_ok=True)
                        shutil.move(str(destination), str(source))
                except Exception as rollback_exc:
                    rollback_errors.append(str(rollback_exc))
            if created_destination and plan.destination_root:
                try:
                    Path(plan.destination_root).rmdir()
                except OSError:
                    pass
        write_journal(
            {
                "id": plan.id,
                "status": "failed",
                "kind": plan.kind,
                "plan": plan.to_dict(),
                "completed": completed,
                "error": str(exc),
                "rollback_errors": rollback_errors,
                "created_destination": created_destination,
            }
        )
        if rollback_errors:
            raise RuntimeError(
                f"{exc}\n자동 롤백 중 확인이 필요한 항목: " + "; ".join(rollback_errors)
            ) from exc
        raise


def _same_fingerprint(path: Path, raw: dict[str, int]) -> bool:
    try:
        return _fingerprint(path) == FileFingerprint(**raw)
    except (OSError, TypeError):
        return False


def _undo_rename(completed: list[dict[str, Any]]) -> None:
    for entry in completed:
        destination = Path(entry["destination"])
        if not destination.is_file() or not _same_fingerprint(
            destination, entry["destination_fingerprint"]
        ):
            raise SafetyError(f"Undo 대상이 변경되었습니다: {destination.name}")

    source_keys = {os.path.normcase(entry["source"]) for entry in completed}
    destination_keys = {os.path.normcase(entry["destination"]) for entry in completed}
    for entry in completed:
        source = Path(entry["source"])
        if source.exists() and os.path.normcase(str(source)) not in destination_keys:
            raise SafetyError(f"원래 파일명이 이미 사용 중입니다: {source.name}")

    parked: list[tuple[Path, Path]] = []
    for entry in completed:
        destination = Path(entry["destination"])
        temp = destination.parent / f".fileflow-undo-{uuid4().hex}.tmp"
        os.replace(destination, temp)
        parked.append((temp, Path(entry["source"])))
    for temp, source in parked:
        os.replace(temp, source)


def undo_latest(progress: Progress | None = None) -> int:
    payload = load_latest()
    if not payload:
        raise SafetyError("실행 취소할 직전 작업이 없습니다.")
    completed = list(payload.get("completed", []))
    if payload.get("kind") == "rename":
        _undo_rename(completed)
    else:
        for entry in completed:
            source = Path(entry["source"])
            destination = Path(entry["destination"])
            if not destination.is_file() or not _same_fingerprint(
                destination, entry["destination_fingerprint"]
            ):
                raise SafetyError(f"Undo 대상이 변경되었습니다: {destination.name}")
            if entry["action"] == "move" and source.exists():
                raise SafetyError(f"원래 위치가 이미 사용 중입니다: {source}")
        for directory in reversed(payload.get("deleted_dirs", [])):
            Path(directory).mkdir(parents=True, exist_ok=True)
        for index, entry in enumerate(reversed(completed), 1):
            source = Path(entry["source"])
            destination = Path(entry["destination"])
            if entry["action"] == "copy":
                destination.unlink()
            elif entry["action"] == "move":
                source.parent.mkdir(parents=True, exist_ok=True)
                shutil.move(str(destination), str(source))
            if progress:
                progress(index, len(completed), source.name)
        if payload.get("created_destination"):
            destination_root = payload.get("plan", {}).get("destination_root")
            if destination_root:
                try:
                    Path(destination_root).rmdir()
                except OSError:
                    pass
    mark_undone(payload)
    return len(completed)
